rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+
+Initially, we will support splitting for the following dataset sources and
+transformations:
+
+*   `batch`, `CsvDataset`, `dense_to_sparse_batch`, `filter`,
+    `FixedLengthRecordDataset`, `flat_map`, `from_tensor_slices`,
+    `group_by_window`, `ignore_errors`, `interleave`, `list_files`, `map`,
+    `range`, `repeat`, `padded_batch`, `prefetch`, `shuffle`, `SSTableDataset`,
+    `TextLineDataset`, `TFRecordDataset`, `unbatch`, `window`.
+
+### Master and worker services
+
+This section discusses the design for the master and worker services. These
+services are used by the Python API to provide distributed dataset processing,
+and these services use the splitting API as a part of their implementation.
+
+#### Master API
+
+The master is responsible for registering datasets, generating and tracking
+iteration and worker ids, and generating dataset splits for processing on
+workers.
+
+Below is a sketch of the Master API. This API is not public and is subject to
+change.
+
+```cpp
+/// ---- Methods called by consumers ----
+
+// Registers a dataset and returns an id for the dataset. If the dataset is
+// already registered, its dataset id is returned.
+int GetOrRegisterDataset(GraphDef dataset);
+
+// Creates and returns `num_consumers` iterator ids which partition the
+// specified dataset. This also creates an internal `iteration_id` used to
+// track the overall dataset iteration. `num_tasks` defines how many tasks to
+// create. If `num_tasks` is -1, it is up to the master to determine how many
+// tasks to create.
+list<int> CreateIterators(int dataset_id, int num_consumers,
+                          int num_tasks);
+
+// Returns the list of tasks processing data for `iterator_id`. Consumers query
+// this to find which worker addresses to read data from.
+list<TaskInfo> GetWorkersForiterator(int iterator_id);
+
+///---- Methods called by input workers ----
+
+// Registers a worker and returns its worker id.
+int RegisterWorker(WorkerInfo worker_info);
+
+// Requests the next splits to process on the given worker for the given
+// iteration_id.
+List<Split> GetSplits(int worker_id, int iteration_id);
+```
+
+#### Worker API
+
+The worker is responsible for processing datasets and providing dataset elements
+to consumers.
+
+Below is a sketch of the Worker API. This API is not public and is subject to
+change.
+
+```cpp
+/// ---- Methods called by consumers ----
+
+// Gets the next element for the specified iterator_id.
+list<Tensors> GetElement(iterator_id);
+
+/// ---- Methods called by master ----
+
+// Requests that the worker process the specified dataset. This will trigger the
+// worker to start requesting splits from the master using the `iteration_id`.
+void ProcessDataset(int dataset_id, int iteration_id, list<int> iterator_ids);
+```
+
+#### Visitation Guarantees
+
+When iterating over a dataset, the tf.data service will process all input data
+at least once, even in the presence of master or worker failures. If there are
```

Comment:
Can you expand a bit on the tradeoffs of not enforcing "exactly once" when failures are present?

Reply:
Revised this section to support exactly once even when failures are present.